// @0xsquid/sdk/dist/types/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CallType = exports.SquidCallType = exports.ChainType = exports.ChainName = void 0;
var ChainName;
(function (ChainName) {
    ChainName["ACRECHAIN"] = "acre";
    ChainName["AGORIC"] = "agoric";
    ChainName["ARBITRUM"] = "Arbitrum";
    ChainName["ARBITRUM2"] = "arbitrum";
    ChainName["ASSETMANTLE"] = "assetmantle";
    ChainName["AURA"] = "aura";
    ChainName["AURORA"] = "aurora";
    ChainName["AVALANCHE"] = "Avalanche";
    ChainName["AXELARNET"] = "axelarnet";
    ChainName["BASE"] = "base";
    ChainName["BINANCE"] = "binance";
    ChainName["CARBON"] = "carbon";
    ChainName["CELO"] = "celo";
    ChainName["COMDEX"] = "comdex";
    ChainName["COSMOS"] = "cosmoshub";
    ChainName["CRESCENT"] = "crescent";
    ChainName["EMONEY"] = "e-money";
    ChainName["ETHEREUM"] = "Ethereum";
    ChainName["ETHEREUM2"] = "Ethereum-2";
    ChainName["EVMOS"] = "evmos";
    ChainName["FANTOM"] = "Fantom";
    ChainName["FETCH"] = "fetch";
    ChainName["FILECOIN"] = "filecoin";
    ChainName["INJECTIVE"] = "injective";
    ChainName["JUNO"] = "juno";
    ChainName["KAVA"] = "kava";
    ChainName["KI"] = "ki";
    ChainName["KUJIRA"] = "kujira";
    ChainName["MOONBEAM"] = "Moonbeam";
    ChainName["NEUTRON"] = "neutron";
    ChainName["OPTIMISM"] = "optimism";
    ChainName["OSMOSIS"] = "osmosis";
    ChainName["OSMOSIS5"] = "osmosis-5";
    ChainName["POLYGON"] = "Polygon";
    ChainName["REGEN"] = "regen";
    ChainName["SEI"] = "sei";
    ChainName["SECRET"] = "secret";
    ChainName["SECRETSNIP"] = "secret-snip";
    ChainName["STARGAZE"] = "stargaze";
    ChainName["STRIDE"] = "stride";
    ChainName["TERRA2"] = "terra-2";
    ChainName["UMEE"] = "umee";
    ChainName["XPLA"] = "xpla";
})(ChainName = exports.ChainName || (exports.ChainName = {}));
var ChainType;
(function (ChainType) {
    ChainType["EVM"] = "evm";
    ChainType["Cosmos"] = "cosmos";
})(ChainType = exports.ChainType || (exports.ChainType = {}));
var SquidCallType;
(function (SquidCallType) {
    SquidCallType[SquidCallType["DEFAULT"] = 0] = "DEFAULT";
    SquidCallType[SquidCallType["FULL_TOKEN_BALANCE"] = 1] = "FULL_TOKEN_BALANCE";
    SquidCallType[SquidCallType["FULL_NATIVE_BALANCE"] = 2] = "FULL_NATIVE_BALANCE";
    SquidCallType[SquidCallType["COLLECT_TOKEN_BALANCE"] = 3] = "COLLECT_TOKEN_BALANCE";
})(SquidCallType = exports.SquidCallType || (exports.SquidCallType = {}));
var CallType;
(function (CallType) {
    CallType["SWAP"] = "SWAP";
    CallType["BRIDGE"] = "BRIDGE";
    CallType["CUSTOM"] = "CUSTOM";
})(CallType = exports.CallType || (exports.CallType = {}));
