// @0xsquid/sdk/dist/error/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SquidError = exports.ErrorType = void 0;
var ErrorType;
(function (ErrorType) {
    ErrorType["UnknownError"] = "UnknownError";
    ErrorType["RouteResponseError"] = "RouteResponseError";
    ErrorType["ValidationError"] = "ValidationError";
    ErrorType["InitError"] = "InitError";
})(ErrorType = exports.ErrorType || (exports.ErrorType = {}));
class SquidError extends Error {
    constructor({ message, errorType, logging, errors, logLevel }) {
        super(message);
        this.logging = false;
        this.message = message;
        this.errorType = errorType;
        this.errors = errors;
        this.logging = !!logging;
        this.logLevel = logLevel;
        Object.setPrototypeOf(this, SquidError.prototype);
        if (this.logging) {
            if (this.logLevel === "error") {
                console.error(`Error type ${errorType}: ${message}`);
            }
            if (this.logLevel === "debug") {
                console.debug(`Error type ${errorType}: ${message}`);
            }
            if (this.logLevel === "info") {
                console.info(`Error type ${errorType}: ${message}`);
            }
        }
    }
}
exports.SquidError = SquidError;
