// @0xsquid/sdk/dist/utils/setAxiosInterceptors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setAxiosInterceptors = void 0;
const error_1 = require("../error");
const setAxiosInterceptors = (axiosInstance, config) => {
    axiosInstance.interceptors.response.use(response => response, error => {
        if (error.response) {
            return Promise.reject(new error_1.SquidError({
                message: error.response.statusText,
                errorType: error.response.data.errors[0].errorType,
                errors: error.response.data.errors,
                logging: config.logging,
                logLevel: config.logLevel
            }));
        }
        return Promise.reject(new error_1.SquidError({
            message: "There was an error while trying to fetch Squid Api",
            errorType: error_1.ErrorType.UnknownError,
            errors: [error],
            logging: config.logging,
            logLevel: config.logLevel
        }));
    });
    return axiosInstance;
};
exports.setAxiosInterceptors = setAxiosInterceptors;
